### PR TITLE
Docker improvements

### PR DESCRIPTION
**Pull request by jessedobbelaere**
## Type
- Enhancement

## Pull request description
Made some adjustments so we can lose some of the xdebug configuration. Also start using .env files so people don't have to modify the docker-compose file to fit their needs.
Tested this with docker-for-mac and XDebug works + symfony commands can also be debugged with xdebug

- [x] Some small adjustments to the Docker files
- [x] Write good documentation about how to setup Docker and how to use it locally
- [x] ~~Add Healthchecks in Dockerfile~~
- [x] Fix Dockerfilelint improvements
- [x] Add docs about setting up phpstorm to run tests and webtestcase tests
- [x] Add "(Failure EADDRINUSE)" to FAQ common errors
- [x] Allow stdout errors to go to docker logs via monolog

### Dockerfilelint

https://github.com/replicatedhq/dockerfilelint

```
npm install -g dockerfilelint
```

<details>

```
❯ dockerfilelint Dockerfile

File:   Dockerfile
Issues: 5

Line 2: MAINTAINER Fork CMS <info@fork-cms.com>
Issue  Category      Title                 Description
    1  Deprecation   Deprecated as of      This INSTRUCTION is deprecated as of Docker 1.13
                     Docker 1.13

Line 8: RUN apt-get update && apt-get install -y \
Issue  Category      Title                 Description
    2  Optimization  apt-get update with   Use of apt-get update should be paired with rm -rf
                     matching cache rm     /var/lib/apt/lists/* in the same layer.
    3  Optimization  Consider              Consider using a `--no-install-recommends` when `apt-get`
                     `--no-install-recomm  installing packages.  This will result in a smaller image size.
                     ends`                 For
                                           more information, see [this blog
                                           post](http://blog.replicated.com/2016/02/05/refactoring-a-dockerfil
                                           e-for-image-size/)

Line 25: RUN apt-get update && apt-get install -y \
Issue  Category      Title                 Description
    4  Optimization  apt-get update with   Use of apt-get update should be paired with rm -rf
                     matching cache rm     /var/lib/apt/lists/* in the same layer.
    5  Optimization  Consider              Consider using a `--no-install-recommends` when `apt-get`
                     `--no-install-recomm  installing packages.  This will result in a smaller image size.
                     ends`                 For
                                           more information, see [this blog
                                           post](http://blog.replicated.com/2016/02/05/refactoring-a-dockerfil
                                           e-for-image-size/)
```

</details>

~~I will fix these issues as well~~ Fixed

### CODE REVIEW

File: .env.dist [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Use this file to set your environment variables for Docker
+
+# XDebug variables
+# Configure the IP of your hostmachine, or use the special Mac-only DNS name which will resolve to the internal IP of the host.
+DOCKER_HOST_IP=docker.for.mac.localhost
+PHP_IDE_CONFIG=serverName=localhost
+XDEBUG_IDEKEY=PHPSTORM
+XDEBUG_PORT=9000
```
Comment (WouterSioen) [inline .env.dist]: this is awesome, not everybody uses phpstorm 👍

File: .htaccess [hunk 2]
```diff
@@ -48,7 +48,7 @@ deny from env=stayout
     RewriteRule src/Backend/Core/Js/ckfinder/core/connector/php/connector\.php - [L]
 
 	# forbidden folders
-    RewriteRule .*\.php - [F]
+        RewriteRule .*\.php - [F]
```
Comment (carakas) [inline .htaccess]: indentation is still wrong

Comment (jessedobbelaere) [inline .htaccess]: Are you sure? Looks ok in the "Files Changed" tab
http://d.pr/i/SlKUK0

**Comment by jessedobbelaere** [comment]
By doing `docker-compose logs -f app` you can now see the php logs as they happen. I also included a formatter with stacktrace enabled.

```
app_1  | 172.18.0.1 - - [23/Aug/2017:02:25:48 +0000] "GET /private/en HTTP/1.1" 302 959 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.101 Safari/537.36"
app_1  | [2017-08-23 02:25:54] request.INFO: Matched route "backend". {"route":"backend","route_parameters":{"_controller":"\\ForkCMS\\App\\ForkController::backendController","_locale":"en","module":"authentication","action":null,"_route":"backend"},"request_uri":"http://localhost/private/en/authentication?querystring=%2Fprivate%2Fen","method":"GET"} []
app_1  | [2017-08-23 02:25:56] request.CRITICAL: Uncaught PHP Exception Exception: "My exception thrown by Jesse :)" at /var/www/html/src/Backend/Core/Engine/Action.php line 48 {"exception":"[object] (Exception(code: 0): Lol jesse at /var/www/html/src/Backend/Core/Engine/Action.php:48)
app_1  | [stacktrace]
app_1  | #0 /var/www/html/src/Backend/Core/Engine/Backend.php(46): Backend\\Core\\Engine\\Action->__construct(Object(ForkCMS\\App\\AppKernel))
app_1  | #1 /var/www/html/app/ForkController.php(111): Backend\\Core\\Engine\\Backend->initialize()
app_1  | #2 /var/www/html/app/ForkController.php(61): ForkCMS\\App\\ForkController->handleApplication(Object(Backend\\Core\\Engine\\Backend))
app_1  | #3 [internal function]: ForkCMS\\App\\ForkController->backendController()
app_1  | #4 /var/www/html/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php(153): call_user_func_array(Array, Array)
app_1  | #5 /var/www/html/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php(68): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
app_1  | #6 /var/www/html/app/Kernel.php(60): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
app_1  | #7 /var/www/html/index.php(66): ForkCMS\\App\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
app_1  | #8 {main}
app_1  | "} []
app_1  | 172.18.0.1 - - [23/Aug/2017:02:25:54 +0000] "GET /private/en/authentication?querystring=%2Fprivate%2Fen HTTP/1.1" 500 153137 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.101 Safari/537.36"
app_1  | [2017-08-23 02:26:02] request.INFO: Matched route "_wdt". {"route":"_wdt","route_parameters":{"_controller":"web_profiler.controller.profiler:toolbarAction","token":"d2c96c","_route":"_wdt"},"request_uri":"http://localhost/_wdt/d2c96c","method":"GET"} []
```

https://github.com/forkcms/forkcms/pull/2168/commits/5c563120b0e4149d7150cc14adb6043c4ed54423

File: .htaccess (hunk 2 above)
Comment (carakas) [inline .htaccess]: Maybe the plus made it look like that
screen shot 2017-08-23 at 09 10 07

**Comment by WouterSioen** [comment]
oops, I tried to do a PR to your branch but github gave me a 500 page first and added the commit to the PR automatically after a refresh. I hope you're ok with my small changes.

**Comment by jessedobbelaere** [comment]
Thanks @WouterSioen !

**Comment by carakas** [comment]
@jessedobbelaere any update on this?

**Comment by WouterSioen** [comment]
this seems almost done looking at the checklist?

**Comment by carakas** [comment]
indeed :) That's why I'm asking :) because I would hate to not ship this when we release 5.1.0

**Comment by jessedobbelaere** [comment]
Ah, yeah i'll write a short phpstorm setup guide (configuring phpstorm for docker with phpunit/xdebug/remote interpreter) and then its good to go I think. Is 5.1.0 released tomorrow?

**Comment by carakas** [comment]
No it isn't, still waiting on some other PR's

**Comment by carakas** [comment]
maybe 5.0.3

**Comment by jessedobbelaere** [comment]
@carakas Think this is good to merge now (after someone proofreads it again maybe?)

File: app/config/config_dev.yml [hunk 3]
```diff
@@ -13,6 +13,12 @@ monolog:
             type:  stream
             path:  "%site.path_www%/var/logs/%kernel.environment%.log"
             level: debug
+        dockerstdout:
+            type: stream
+            path: php://stdout
+            level: debug
+            formatter: monolog.full_trace_formatter
+            channels: [!event]
```
Comment (WouterSioen) [inline app/config/config_dev.yml]: this also works without docker?

File: docs/01. installation/07. docker.md [hunk 4]
```diff
@@ -1,37 +1,352 @@
...
+* [PHP 7.1](https://store.docker.com/images/php) (with extensions GD2, pdo, zip, intl, ... and XDebug)
 * Apache
-* MariaDB
+* [MariaDB](https://store.docker.com/images/mariadb) (latest)
 
 **Features:**
-* Easily switch between PHP versions: 7.1, 7.0, ...
+* Easy to set and start up a project (`docker-compose up`).
+* Easily disposable stack (`docker-compose down`).
+* Easily switch between PHP versions: 7.1, 7.2, ... by editting the Dockerfile to fit your needs.
```
Comment (WouterSioen) [inline docs/01. installation/07. docker.md]: editing (no double t)

File: docs/01. installation/07. docker.md [hunk 5]
```diff
@@ -124,61 +146,93 @@ You can easily use different credentials by adjusting them in the `docker-compos
...
+
+- We also need to configure PHPStorm to use the PHP interpreter that resides inside the app container. To do so, go to 
+preferences to "Languages & Frameworks > PHP". Click the "..." icon for the PHP interpreter. Create a new interpreter
+and use type "Docker". Select the Docker for mac server you created, and the Docker image name for the project you're working on.
+Make sure PHPStorm detects your php version correctly (see video above). Next, configure the Docker container by clicking the "...",
+and fill in the Network mode with the value you find when running `docker network ls`. Next, go to Volume bindings and
+make sure the container path is set to `/var/www/html`. This is the path where our code lives inside the container. 
+
+- Let's configure Xdebug now. In PHPStorm's preferences, go to the PHP > Debug section. Make sure Xdebug accepts connections and 
+listens on port 9000 (should be default configuration). ![Configuring XDebug in PHPStorm](assets/phpstorm-xdebug1.png)
```
Comment (WouterSioen) [inline docs/01. installation/07. docker.md]: @forkcms/core-contributors can one of the PHPStorm users follow these instructions and confirm that it's working and easy to follow?

Seems clear to me at first sight though.

Comment (jessedobbelaere) [reply docs/01. installation/07. docker.md]: 👍  easiest would be to follow along with the video. Let me know any remarks while setting it up

File: app/config/config_dev.yml (hunk 3 above)
Comment (jessedobbelaere) [reply app/config/config_dev.yml]: Yeah I guess, only in dev mode. Do you mean running commands other than in docker could be affected?

**Pull request merged**